Use model classes instead of CIMCORE json to construct components.

mint-thompson opened the pull request:
When making constraints and elements to put into the modeldoc, use the model classes from shr-models
in their original forms. This results in some changes to attribute paths.
When an element definition contains a constraint on a field's subfield, and that field is declared by this element (not inherited),
show those constraints in the "Properties Constrained" portion of the model doc.

This pull request is intended to be applied along with accompanying pull requests on shr-cli and shr-models:
* https://github.com/standardhealth/shr-models/pull/48
* https://github.com/standardhealth/shr-cli/pull/226

mint-thompson committed:
Use model classes instead of CIMCORE json to construct components.

When making constraints and elements to put into the modeldoc, use the model classes from shr-models
in their original forms. This results in some changes to attribute paths.
When an element definition contains a constraint on a field's subfield, and that field is declared by this element (not inherited),
show those constraints in the "Properties Constrained" portion of the model doc.

cmoesel commented on components/constraints.js:
```diff
@@ -10,9 +10,9 @@ function cardToString(card) {
   let max = '*';
   if (card) {
-    if ('min' in card)
+    if (typeof card.min != 'undefined')
```
In cases like this, I generally prefer
```js
if (card.min != null)
```
It's shorter and easier to read, but it also tests against both `null` and `undefined` -- which is what I think we intended in this case.  You'll still find the `typeof x === 'undefined'` construct in SHR code here or there, but usually that's from _before_ I realized the benefit of `== null`/`!= null`.  (Note -- double _not triple_ equals).

cmoesel replied:
This shows up below as well, but I won't call it out again.

cmoesel commented on components/constraints.js:
```diff
@@ -96,139 +96,115 @@ class Constraints {
-      const iConstraint = this.newConstraint(name, value, subpath, lastMod);
-      this.constraints.push(iConstraint);
-    });
+    const system = constraint.code.system ? constraint.code.system : '';
```
The above is another construct you probably saw in other SHR code (probably written by me!), but I think can be simplified to:
```js
const system = constraint.code.system || '';
```

cmoesel commented on components/constraints.js:
```diff
@@ -96,139 +96,115 @@ class Constraints {
+  buildFullPath(constraint) {
+    if(constraint.path.length > 0) {
+      return constraint.path.reduce((fullPath, pathPiece) => {
+        if(pathPiece.fqn in this.elements) {
```
Are primitives in `this.elements`?  I feel like they're not, but I could be wrong (I don't have much experience in this module).  If primitives are not in `this.elements` then are we dropping them when building the full path like this?

mint-thompson replied:
`this.elements` does not contain primitives, so you are correct in identifying that this code results in the primitive's name not being part of the path. This is consistent with the module's previous behavior.
Given that the intent is to avoid adding the primitive's name to the path, this intent is clearer by directly checking `pathPiece.namespace == 'primitive'`. Does that sound reasonable to you?

cmoesel commented on components/constraints.js:
```diff
@@ -239,18 +215,42 @@ class Constraints {
-      Object.keys(this.field.constraints).forEach((cType) => {
-        const constraint = this.field.constraints[cType];
-        this.switchConstraintType(constraint, cType, this.field.name);
+    if ('constraints' in this.field && 'constraintsFilter' in this.field) {
```
Trivial thing, but I always thought it was kind of weird the way Abhijay did this.  Maybe we should go ahead and use the more traditional:
```js
if (this.field.constraints && this.field.constraintsFilter) {
```

cmoesel reviewed (changes requested):
The code looks really good -- and kudos for your attention to all the little details in making this work (there are a lot!).  I put a few comments in on minor adjustments to consider.

I tried to run this, but was unable to.  Here is my branch setup:
```
shr-cli                  [remove-cimcore-export]
shr-data-dict-export     [master]
shr-es6-export           [master]
shr-expand               [master]
shr-fhir-export          [master]
shr-grammar              [master]
shr-json-javadoc         [remove-cimcore-export]
shr-json-schema-export   [master]
shr-models               [remove-cimcore-export]
shr-test-helpers         [master]
shr-text-import          [master]
shr_spec                 [master]
```
I ran `./yarn-link-all` and then executed the CLI and got this error:
```
/Users/cmoesel/dev/standardhealth/shr-cli/app.js:93
const errorFiles = [shrTI.errorFilePath(), shrEx.errorFilePath(), shrFE.errorFilePath(), shrJDE.errorFilePath(),
                                                                                                ^

TypeError: shrJDE.errorFilePath is not a function
    at Object.<anonymous> (/Users/cmoesel/dev/standardhealth/shr-cli/app.js:93:97)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:643:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:839:10)
    at internal/main/run_main_module.js:17:11
```

I think that Nick's new logger refactoring snuck in on master under you and you need to rebase to pull in his changes.  Once you've done that, let me know and I'll give it a try again.

Feel free to reach out if you have any questions!

mint-thompson committed:
Merge branch 'master' into remove-cimcore-export

# Conflicts:
#	components/constraints.js
#	export.js

mint-thompson committed:
Improve readability of conditionals

Compare with null instead of checking a variable's type is undefined when making cardinality string.
Use simpler conditional when getting code system for "includes code" constraint.
Check value's truthiness instead of using "in" operator when checking if a field has constraints and a constraintsFilter.
Since the full path should not include the names of primitives, directly check if the path piece's namespace is "primitive".

cmoesel reviewed (approved):
Nice job, @mint-thompson.  Code looks good and modeldoc looks right.  Did a diff on modeldoc outputs and selected random element pages -- and all diffs made sense.  Couldn't do it for every single one (lots of whitespace-only changes muddying the waters), but based on random spot-checks, looks great!
